hasjob/views/helper.py

```diff
@@ -320,9 +320,68 @@ def bgroup(jobpost_ab, post):
+def get_jobpost_impressions(jobpost_id):
+    return db.session.query(db.func.count(db.func.distinct(EventSession.user_id)).label('count')).join(
+        JobImpression).filter(JobImpression.jobpost_id == jobpost_id).first().count
+
+
+def get_post_viewcounts(jobpost_id):
+    cache_key = JobPost.viewcounts_key(jobpost_id)
+    values = g.viewcounts.get(cache_key)
+    jobpost = None
+    if not values:
+        values = redis_store.hgetall(cache_key)
+        jobpost = JobPost.get_by_id(jobpost_id)
+    if 'impressions' not in values:
+        jobpost = jobpost or JobPost.get_by_id(jobpost_id)
+        values['impressions'] = get_jobpost_impressions(jobpost.id)
+        redis_store.hset(cache_key, 'impressions', values['impressions'])
+        redis_store.expire(cache_key, 86400)
+    else:
+        values['impressions'] = int(values['impressions'])
+    if 'viewed' not in values:
+        # values['viewed'] = UserJobView.query.filter_by(jobpost=jobpost).count()
+        jobpost = jobpost or JobPost.get_by_id(jobpost_id)
+        values['viewed'] = jobpost.viewcounts_viewed
+        redis_store.hset(cache_key, 'viewed', values['viewed'])
+        redis_store.expire(cache_key, 86400)
+    else:
+        values['viewed'] = int(values['viewed'])
+    if 'opened' not in values:
+        # values['opened'] = UserJobView.query.filter_by(jobpost=jobpost, applied=True).count()
+        jobpost = jobpost or JobPost.get_by_id(jobpost_id)
+        values['opened'] = jobpost.viewcounts_opened
+        redis_store.hset(cache_key, 'opened', values['opened'])
+        redis_store.expire(cache_key, 86400)
+    else:
+        values['opened'] = int(values['opened'])
+    if 'applied' not in values:
+        # values['applied'] = JobApplication.query.filter_by(jobpost=jobpost).count()
+        jobpost = jobpost or JobPost.get_by_id(jobpost_id)
+        values['applied'] = jobpost.viewcounts_applied
+        redis_store.hset(cache_key, 'applied', values['applied'])
+        redis_store.expire(cache_key, 86400)
+    else:
+        values['applied'] = int(values['applied'])
+    # pay_label rendering is extraordinarily slow. We don't know why yet, but it's static data, so cache it
+    if 'pay_label' not in values:
+        jobpost = jobpost or JobPost.get_by_id(jobpost_id)
+        values['pay_label'] = jobpost.pay_label()
+        redis_store.hset(cache_key, 'pay_label', values['pay_label'].encode('utf-8'))
+        redis_store.expire(cache_key, 86400)
+    elif isinstance(values['pay_label'], str):  # Redis appears to return bytestrings, not Unicode
+        values['pay_label'] = unicode(values['pay_label'], 'utf-8')
+    return values
+
+
+@app.context_processor
+def inject_post_viewcounts():
+    return {'get_post_viewcounts': get_post_viewcounts}
+
+
+def load_viewcounts(posts):
     redis_pipe = redis_store.pipeline()
-    viewcounts_keys = [p.viewcounts_key for p in posts]
+    viewcounts_keys = [JobPost.viewcounts_key(p.id) for p in posts]
```

Comment:
This is ~700 calls in a row in production. That slowdown begins to be noticeable. Make `viewcounts_key` accept a list and return a list (test with `isinstance(jobpost_id, (list, tuple))`)